### PR TITLE
Staff crud backend

### PR DESCRIPTION
**Pull request by GeorgebigG**
I finished the back-end code for the Staff Crud Epic. I added get, update, delete, get all and create functions:

I added:
api/stuff/update
api/stuff/post
api/stuff/get
api/stuff/all
api/stuff/delete

and I created a test file that tested all of these features:

https://organic-george.dokku-09.cs.ucsb.edu

### CODE REVIEW

**Review by Peak1260 — DISMISSED** [review]
LGTM

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    public Iterable<Staff> allCourses() {
+        User u = getCurrentUser().getUser();
+        log.info("u={}", u);
+
+        return staffRepository.findAll();
+    }
+
+    @Operation(summary = "Create a new staff")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN')")
+    @PostMapping("/post")
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: You can remove post and just make this ("").

It is better api design to let the HTTP verb (GET,POST,PUT,DELETE) do the work.

adding /post is redundant.

**Review by pconrad — CHANGES_REQUESTED** [review]
Please simplify the URLs.

It's better to just use "" when the HTTP method (GET, POST, DELETE, PUT) disambiguates the request.

**Review by pconrad — CHANGES_REQUESTED** [review]
dokku deployment gives me:

image

please get it working

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java (hunk 1 above)
Comment (GeorgebigG) [reply src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: Thank you for letting me know! I fixed it

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        return staff;
+    }
+
+    @Operation(summary = "Update information for a staff member")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN')")
+    @PutMapping("")
+    public Staff updateStaff(
+            @Parameter(name = "id") @RequestParam Long id,
+            @Parameter(name = "courseId", description = "course id") @RequestParam Long courseId,
+            @Parameter(name = "githubId", description = "school abbreviation e.g. UCSB") @RequestParam Integer githubId)
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]:
```suggestion
            @Parameter(name = "githubId", description = "github integer login id (not github username)") @RequestParam Integer githubId)
```

Comment (pconrad) [reply src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: This is incorrect on Swagger, and it does make a difference.

image

**Review by pconrad — CHANGES_REQUESTED** [review]
What is the `user` field here and why is it `null`?

image

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    @Operation(summary = "Create a new staff")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN')")
+    @PostMapping("")
+    public Staff postCourse(
+            @Parameter(name = "courseId", description = "course id") @RequestParam Long courseId,
+            @Parameter(name = "githubId", description = "school abbreviation e.g. UCSB") @RequestParam Integer githubId)
+            throws JsonProcessingException {
+
+
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java:60]: You need to add a check here that checks whether the githubId exists in the Users table.  If not, you need to return an error.  See, for example, this code in the `CoursesController`.  This uses the `Course` table to look up a course, but it's the same idea; you'll just be looking up a `User` by `githubId`:

```
   Course course = courseRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(Course.class, id.toString()));
```

Comment (pconrad) [reply src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java:60]: And for that matter, you also need to be validating the course id, which you can do with *exactly* this code.

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        return staffRepository.findAll();
+    }
+
+    @Operation(summary = "Create a new staff")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN')")
+    @PostMapping("")
+    public Staff postCourse(
+            @Parameter(name = "courseId", description = "course id") @RequestParam Long courseId,
+            @Parameter(name = "githubId", description = "school abbreviation e.g. UCSB") @RequestParam Integer githubId)
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]:
```suggestion
           @Parameter(name = "githubId", description = "github integer login id (not github username)") @RequestParam Integer githubId)
```

Your description is incorrect (copy/pasted code?)...

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    public Staff updateStaff(
+            @Parameter(name = "id") @RequestParam Long id,
+            @Parameter(name = "courseId", description = "course id") @RequestParam Long courseId,
+            @Parameter(name = "githubId", description = "school abbreviation e.g. UCSB") @RequestParam Integer githubId)
+            throws JsonProcessingException {
+
+        Staff staff = staffRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(Staff.class, id.toString()));
+
+        staff.setCourseId(courseId);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: No... don't allow editing of these fields.  The only thing we should be doing here is allowing a user to delete a staff member.  If you are changing the github id or the course id, that should be a whole other record.  The reason is that staff ids might be associated with the audit trail of what someone did in the app. You would not want an admin to change the github id associated with a staff member from Alice to Bob and suddenly Bob is responsible for everythign Alice may have done.

This is another reason to put these operations in as separate PRs.

Comment (pconrad) [reply src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: Basically: just remove PUT / Edit completely.  We will only have POST, GET and DELETE for staff members (for now).

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    public Staff deleteCourse(
+            @Parameter(name = "id") @RequestParam Long id)
+            throws JsonProcessingException {
+
+        Staff staff = staffRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(Staff.class, id.toString()));
+
+        // Check if the current user is a staff member for this course or an admin. If
+        // not, throw AccessDeniedException
+
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: You have the comment but no matching code.   Either put in the code (you should!) or delete the comment (but you should put in the code.  Another reason to make this a separate PR.)

**Review by pconrad — CHANGES_REQUESTED** [review]
Several changes to request.

I also want to STRONGLY encourage you to break this into multiple PRs, if for no other reason, because you'll earn 30 points for the team that way instead of 10, and your team *really, really* needs the points right now.

But it's up to you.

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java (hunk 3 above)
Comment (GeorgebigG) [reply src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java:60]: Fixed it!

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java (hunk 6 above)
Comment (GeorgebigG) [reply src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: Fixed

File: src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java (hunk 5 above)
Comment (GeorgebigG) [reply src/main/java/edu/ucsb/cs156/organic/controllers/StaffController.java]: Ok i removed the Edit and modified the test cases

**Comment by GeorgebigG** [comment]
Ok i will split branches into:
staff-crud-backend-post
staff-crud-backend-get
staff-crud-backend-all
staff-crud-backend-delete

To help my team get more points

**Comment by GeorgebigG** [comment]
It sucks but unfortunately i added all of these features: 
api/stuff/update
api/stuff/post
api/stuff/get
api/stuff/all
api/stuff/delete

but we didn't end up using them so 500 lines of code basically went to waste.
